paint/data/stac_client.py

```diff
@@ -0,0 +1,560 @@
+        and raise a value error indicating that the desired catalog does not exist.
+
+        Parameters
+        ----------
+        href : str
+            URL reference to the catalog that should be accessed.
+
+        Returns
+        -------
+        pystac.catalog.Catalog
+            Catalog, if it exists.
+
+        Raises
+        ------
+        ValueError
+            If the URL does not exist.
+        """
+        try:
+            return pystac.Catalog.from_file(href=href)
+        except Exception as e:
+            # Log the error message for inspection.
+            error_message = str(e)
+            # Handle the 404 error explicitly by checking the message content.
+            if "404" in error_message or "Could not read uri" in error_message:
+                raise ValueError(
+                    f"The desired catalog with the URL '{href}' doesn't exist (404 error)!"
+                )
+            else:
+                # Raise the original exception for all other errors
+                raise
+
+    @staticmethod
+    def get_child(
+        parent: Union[pystac.catalog.Catalog, pystac.collection.Collection],
+        child_id: str,
+    ) -> Union[pystac.catalog.Catalog, pystac.collection.Collection, pystac.item.Item]:
+        """
+        Help to get a child from a STAC catalog or collection.
+
+        Since pystac returns `None` if a child is not present, it is important we log a warning if this is the case.
+
+        Parameters
+        ----------
+        parent : Union[pystac.Catalog, pystac.Collection]
+            Parent STAC catalog or collection.
+        child_id : str
+            ID of the child STAC catalog, collection, or item.
+
+        Returns
+        -------
+        Union[pystac.Catalog, pystac.Collection, pystac.Item]
+            Child STAC catalog, collection, or item.
+        """
+        child = parent.get_child(child_id)
+        if child is None:
+            log.warning(
+                f"The child with ID {child_id} is not available, data for this child will not be downloaded."
+            )
+        return child
+
+    def download_file(self, url: str, file_name: Union[str, pathlib.Path]) -> None:
+        """
+        Download a file.
+
+        Parameters
+        ----------
+        url : str
+            URL of the file to download.
+        file_name : Union[str, pathlib.Path]
+            File name to be saved.
+        """
+        response = requests.get(url, stream=True)
+
+        # Open the file in binary write mode and download the content
+        with open(file_name, "wb") as file:
+            downloaded_length = 0
+            for data in response.iter_content(chunk_size=self.chunk_size):
+                file.write(data)
+                downloaded_length += len(data)
+
+    def process_heliostat_items(
+        self,
+        items: list[pystac.item.Item],
+        start_date: Optional[datetime],
+        end_date: Optional[datetime],
```

Comment:
Do we want to use `Optional[...]` or `Union[... | None] = None` (Python3.9) / `... | None = None` (Pythonv>=3.10)? I think @Markus-Goetz had a rather strong opinion on this one.

Reply:
I have switched everything to `Union[ ...., None]` in the STAC client